src/libs/elektra/split.c

```diff
@@ -168,6 +184,12 @@ ssize_t splitAppend (Split * split, Backend * backend, Key * parentKey, int sync
 	split->parents[n] = parentKey;
 	split->syncbits[n] = syncbits;
 
+	// FIXME: somehow find correct sizes
+	split->systemsizes[n] = 0;
+	split->usersizes[n] = 0;
+	split->specsizes[n] = 0;
+	split->dirsizes[n] = 0;
```

Comment:
Theses values were part of the `Backend` struct passed into this function before. I don't know what these sizes are used for, so I am not sure how to solve this now.

Reply:
In the previous implementation there were similar sizes (the names were singular, though) in Split which were used to track how many keys in the respective namespace exist.

Reply:
Like I said, previously these values where stored *within* the `Backend` struct. Therefore, they where automatically passed around.

In `splitBuildup` the `mountGetBackend` function used to return a `Backend *` containing these values. This pointer was then passed to `splitAppend` and simply stored in the `Split` struct.

However, now `mountGetBackend` returns a `Plugin *`, meaning the sizes are not available inside `splitAppend`.

In general, I am not sure we even need these fields at all. If they are supposed to store the number of keys in the respective namespace, then all except the one matching the namespace of `split->parent[n]` should be zero. The one with the matching namespace should just be `ksGetSize (split->keysets[n])`, if I understand the concept correctly.

Reply:
Without checking the code: The different sizes for different namespaces are needed in the Backend struct as cascading backends might have keys across namespaces. And this is probably also the situation where they differ between the Backend and Split struct.

Reply:
I tried to get around this in the cache implementation (using `ksGetSize()`) and failed. I didn't want to dive too deep into the split code, so now the cache simply persists these values.

@markus2330 could the values of `split->systemsizes[n]` etc. be reconstructed somehow? It would be nice to get rid of all those `split->zzzsizes[n]` members.

Reply:
With the current logic (making `kdbSet` a NOP for the backends where the keys are not changed) the sizes are needed somewhere. We use the sizes to check if the same amount of unsynced keys would be written (NOP situation). To have the sizes both per Backend and Split, depending on the implementation, might be unnecessary. It might be, however, a major rewrite with little benefits. (iirc there are many tests that check these sizes.)

Having new higher-level tests that directly check the outcome of the whole `kdbSet` (which plugins are called with which keys) would be very beneficial. @PhilippGackstatter @mpranj: What do you think?

Reply:
Sounds like it makes no sense to rewrite the `split->zzzsizes[n]` code then.

Higher level tests make sense to me. Being able to track which plugins are called with which keys would be great. I needed this quite extensively myself when debugging problems.

Reply:
Having these kinds of tests sounds good from my end. It would probably be benficial for a reimplementation. It also sounds like a lot of work though, so it needs to make sense from other points of views as well.

Reply:
I might quite sure I understand everything, but to me it sounds like these `size` variables do in fact have a use. However, it also sounds like the could be maintained within the `backend` plugin itself. The NOP case could then be detected within the `elektraBackendSet`, which would return ASAP.

Reply:
I may have found a way of avoiding this problem by rethinking splits entirely. See #3492.